ta/sks/src/entry.c

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018-2019, Linaro Limited
+ */
+
+#include <compiler.h>
+#include <sks_ta.h>
+#include <tee_internal_api.h>
+#include <tee_internal_api_extensions.h>
+
+TEE_Result TA_CreateEntryPoint(void)
+{
+	return TEE_SUCCESS;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t __unused param_types,
+				    TEE_Param __unused params[4],
+				    void **session)
+{
+	*session = NULL;
+
+	return TEE_SUCCESS;
+}
+
+void TA_CloseSessionEntryPoint(void *session __unused)
+{
+}
+
+/*
+ * Entry point for SKS TA commands
+ *
+ * ABI: param#0 is the input control buffer with serialized arguments.
+ *	param#1 is an input/output data buffer
+ *	param#2 is an input/output data buffer (also used to return handles)
+ *	param#3 is not used
+ */
+TEE_Result TA_InvokeCommandEntryPoint(void *tee_session __unused, uint32_t cmd,
+				      uint32_t ptypes,
+				      TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Param *ctrl __maybe_unused = NULL;
```

Comment:
This could be simplified by having parameter 0 inout always and parameters 1 and 2 inout if provided.

Reply:
For param#0, yes we can easily expect it INOUT.

Param#1 is used to provide input data (ciphering, signing, ...). Using INOUT will lead to useless memory copy of input data as output data. Few exception command may use it as output, but there's currently none in my sks/pkcs#11 branch. I will remove INOUT case and expect only INPUT.

Param#2 is mostly used to provide output data. If client using an INOUT memref, there will use useless memory copy before TEE is invoked. In my current SKS/pkcs#11 branch, there is only 1 command that uses param#2 as INPUT (command `C_Verify()` expects needs 2 input data buffer).
For this param#2 I would rather allow client some flexibility.

Reply:
Fair enough, but I'd like to have checks that param#2 has the correct IN/OUT/INOUT direction for the function in question.

Reply:
Ok, I'll push a fixup change to be more strict here with expected IN/OUT/INOUT direction.

The more complete TA I have does this the following way:
This entry function checks the params type and passes the `TEE_Param` reference to a function dedicated to the target command (example [here for opening a PKCS#11 session](https://github.com/etienne-lms/optee_os/blob/sks/ta/secure_key_services/src/entry.c#L198)). This handler function check that params references are the expected one: NULL or valid memref, checks also its size of course (example [here](https://github.com/etienne-lms/optee_os/blob/sks/ta/secure_key_services/src/pkcs11_token.c#L949)).